### PR TITLE
Introduce different evaluation types for different binding fields

### LINKED ISSUE #2862: [Bug] Modify dynamic binding evaluation to support escaped characters without double escapes
Escape characters (like `\t`, `\n` etc) are currently not read as such during evaluation. They need to be written as a double escaped `\\n` for the character to make sense while evaluation. This is non-intuitive since the editor space is not represented as a string literal.

### LINKED ISSUE #3759: Introduce different evaluation substitution types
Today we only substitute bindings within a string/JSON via template substitution. We should introduce different substitution types in evaluation to handle scenarios in the API and query pane

- Template substitution: for the final output being templated string with binding evaluations
- Parameter substitution: for the final output be a parameterised string with `$1`, `$2` like prepared statements
- Smart substitution: for JSON like outputs where the value of evaluation decides how it will be substituted in the whole JSON string

### LINKED ISSUE #3846: [Bug] Loading an action leads to an update which wipes out it's dynamicBindingPathList
## Description
There are some unnecessary updates to action when we switch to its pane, removing the dynamicBindingPathList. This causes an issue in the evaluation of the action field.

### Steps to reproduce the behaviour:
1. Go to an API, add some bindings in the header
2. refresh the page
3. Focus on the header field
4. The value is not evaluated

### PR DESCRIPTION
**Pull request by hetunandu**
## Description
Today we only substitute bindings within a string/JSON via template substitution. This PR introduces different substitution types in evaluation to handle scenarios in the API and query pane

- Template substitution: for the final output being templated string with binding evaluations
- Parameter substitution: for the final output be a parameterised string with $1, $2 like prepared statements
- Smart substitution: for JSON like outputs where the value of evaluation decides how it will be substituted in the whole JSON string

Fixes #3759
Fixes #3846
Fixes #2862

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Jest

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/client/src/entities/Widget/utils.ts [hunk 1]
```diff
@@ -116,13 +124,13 @@ export const getAllPathsFromPropertyConfig = (
 };
 
 export const nextAvailableRowInContainer = (
-  parenContainertId: string,
+  parenContainerId: string,
```
Comment (hetunandu) [inline app/client/src/entities/Widget/utils.ts:127]:
```suggestion
  parentContainerId: string,
```

File: app/client/src/entities/Widget/utils.ts [hunk 2]
```diff
@@ -116,13 +124,13 @@ export const getAllPathsFromPropertyConfig = (
...
   canvasWidgets: { [widgetId: string]: FlattenedWidgetProps },
 ) => {
   return (
     Object.values(canvasWidgets).reduce(
       (prev: number, next: any) =>
-        next?.parentId === parenContainertId && next.bottomRow > prev
+        next?.parentId === parenContainerId && next.bottomRow > prev
```
Comment (hetunandu) [inline app/client/src/entities/Widget/utils.ts:133]:
```suggestion
        next?.parentId === parentContainerId && next.bottomRow > prev
```

File: app/client/src/constants/AppsmithActionConstants/formConfig/ApiEditorConfigs.ts [hunk 3]
```diff
@@ -8,10 +8,25 @@ export default [
         configProperty: "actionConfiguration.path",
         controlType: "QUERY_DYNAMIC_INPUT_TEXT",
       },
+      {
+        label: "Body",
```
Comment (trishaanand) [inline app/client/src/constants/AppsmithActionConstants/formConfig/ApiEditorConfigs.ts:12]: Don't we want to introduce another field in this json which dictates the evaluation function used?

Comment (hetunandu) [reply app/client/src/constants/AppsmithActionConstants/formConfig/ApiEditorConfigs.ts:12]: How so? The control type can decide the evaluation function, right?

Comment (mohanarpit) [reply app/client/src/constants/AppsmithActionConstants/formConfig/ApiEditorConfigs.ts:12]: In this case, I think a separate field is better. For example, we want to have a SQL body field without smart substitution and one with smart substitution. We can re-use the same `controlType` for both the fields with different evaluators. This also makes it the distinction clear between the evaluated value pane & the text field itself.

**Review by trishaanand — DISMISSED** [review]

**Review by trishaanand — DISMISSED** [review]

**Review by trishaanand — APPROVED** [review]

**Review by trishaanand — APPROVED** [review]

**Pull request merged**